Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp

```diff
@@ -1010,35 +1000,114 @@ void GraphicsContextSkia::drawSkiaPattern(const sk_sp<SkImage>& image, const Int
+            auto picture = recorder.finishRecordingAsPicture();
+            paint.setShader(picture->makeShader(SkTileMode::kRepeat, SkTileMode::kRepeat, samplingOptions.filter, &shaderMatrix, nullptr));
+        }
     }
 
     m_canvas.drawRect(destRect, paint);
 }
 
-void GraphicsContextSkia::drawPattern(NativeImage& nativeImage, const FloatRect& destRect, const FloatRect& tileRect, const AffineTransform& patternTransform, const FloatPoint& phase, const FloatSize& spacing, ImagePaintingOptions options)
+HashMap<const SkImage*, std::unique_ptr<GLFence>> GraphicsContextSkia::releaseImageFences()
 {
-    auto* nativeImageGrContext = nativeImage.backend().skiaGrContext();
-    if (nativeImageGrContext && nativeImageGrContext != PlatformDisplay::sharedDisplay().skiaGrContext()) {
-        // If we encounter an accelerated NativeImage (skiaGrContext() != nullptr), we are in the threaded GPU rendering painting mode, verify that.
-        ASSERT(m_renderingMode == RenderingMode::Accelerated);
-        ASSERT(ProcessCapabilities::canUseAcceleratedBuffers());
-#if USE(COORDINATED_GRAPHICS)
-        ASSERT(SkiaPaintingEngine::numberOfGPUPaintingThreads() > 0);
-#endif
-        // The 'nativeImage' was produced on another thread -- to use it here, we need to create a new NativeImage, that wraps the existing GPU resource.
-        if (auto newNativeImage = nativeImage.backend().copyAcceleratedNativeImageBorrowingBackendTexture())
-            drawSkiaPattern(newNativeImage->platformImage(), newNativeImage->size(), destRect, tileRect, patternTransform, phase, spacing, options);
-        return;
+    return WTFMove(m_imageFences);
+}
+
+std::unique_ptr<GLFence> GraphicsContextSkia::createAcceleratedRenderingFence(const sk_sp<SkImage>& image)
+{
+    if (!image)
+        return nullptr;
+
+    auto* glContext = PlatformDisplay::sharedDisplay().skiaGLContext();
+    if (!glContext || !glContext->makeContextCurrent())
+        return nullptr;
+
+    auto* grContext = PlatformDisplay::sharedDisplay().skiaGrContext();
+    RELEASE_ASSERT(grContext);
+
+    grContext->flush(image);
+
+    if (GLFence::isSupported()) {
+        grContext->submit(GrSyncCpu::kNo);
+
+        if (auto fence = GLFence::create())
+            return fence;
     }
 
-    drawSkiaPattern(nativeImage.platformImage(), nativeImage.size(), destRect, tileRect, patternTransform, phase, spacing, options);
+    grContext->submit(GrSyncCpu::kYes);
+    return nullptr;
+}
+
+std::unique_ptr<GLFence> GraphicsContextSkia::createAcceleratedRenderingFence(SkSurface* surface)
+{
+    if (!surface)
+        return nullptr;
+
+    auto* glContext = PlatformDisplay::sharedDisplay().skiaGLContext();
+    if (!glContext || !glContext->makeContextCurrent())
+        return nullptr;
+
+    auto* grContext = PlatformDisplay::sharedDisplay().skiaGrContext();
+    RELEASE_ASSERT(grContext);
+
+    grContext->flush(surface);
```

Comment:
And here. Also the code is the same, I'm not an expert on c++ templates, but I think we could use just one implementation using a template, no?

Reply:
The SkSurface variant is not used in this patch, but in the one converting ImageBitmap. Need to think about it.

Reply:
I will postpone the introduction of the SkSurface variant for a follow-up PR.